tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    let histo = Arc::new(Mutex::new(RotatingHistogram::new(period)));
+    service_with_histo(service, policy, min_data_points, latency_percentile, histo)
+}
+
+/// A hedge middleware with a prepopulated latency histogram.  This is usedful
+/// for integration tests.
+pub fn service_with_mock_latencies<S, P, Request>(
+    service: S,
+    policy: P,
+    min_data_points: u64,
+    latency_percentile: f32,
+    period: Duration,
+    latencies_ms: &[u64],
+) -> Service<S, P>
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    let histo = Arc::new(Mutex::new(RotatingHistogram::new(period)));
+    {
+        let mut locked = histo.lock().unwrap();
+        for latency in latencies_ms.iter() {
+            locked.read().record(*latency).unwrap();
+        }
+    }
+    service_with_histo(service, policy, min_data_points, latency_percentile, histo)
+}
+
+fn service_with_histo<S, P, Request>(
+    service: S,
+    policy: P,
+    min_data_points: u64,
+    latency_percentile: f32,
+    histo: Histo,
+) -> Service<S, P>
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    // Clone the underlying service and wrap both copies in a middleware that
+    // records the latencies in a rotating histogram.
+    let recorded_a = Latency::new(histo.clone(), service.clone());
+    let recorded_b = Latency::new(histo.clone(), service);
+
+    // Check policy to see if the hedge request should be issued.
+    let filtered = Filter::new(recorded_b, PolicyPredicate(policy.clone()));
+
+    // Delay the second request by a percentile of the recorded request latency
+    // histogram.
+    let delay_policy = DelayPolicy {
+        histo: histo.clone(),
+        latency_percentile,
+    };
+    let delayed = Delay::new(delay_policy, filtered);
+
+    // If the request is retryable, issue two requests -- the second one delayed
+    // by a latency percentile.  Use the first result to complete.
+    let select_policy = SelectPolicy {
+        policy,
+        histo,
+        min_data_points,
+    };
+    Select::new(select_policy, recorded_a, delayed)
+}
+
+// TODO: Remove when Duration::as_millis() becomes stable.
+fn duration_as_millis(d: Duration) -> u64 {
+    d.as_secs() * 1_000 + d.subsec_millis() as u64
+}
+
+impl latency::Record for Histo {
+    fn record(&mut self, latency: Duration) {
+        let mut locked = self.lock().unwrap();
+        locked
+            .write()
+            .record(duration_as_millis(latency))
+            .unwrap_or_else(|e| {
+                error!("Failed to write to hedge histogram: {:?}", e);
+            })
+    }
+}
+
+impl<P, Request> tower_filter::Predicate<Request> for PolicyPredicate<P>
+where
+    P: Policy<Request>,
+{
+    type Future = future::Either<
```

Comment:
I've been thinking that `Predicate` should not be async. There is no use case yet. This would simplify things here. Thoughts?

Reply:
Yeah, this being non-async would be more convenient.  An async version could be useful for something that makes calls out to an external policy service perhaps... but maybe that could be a different middleware.